### PR TITLE
[move-vm] Add strict/paranoid runtime verification

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation

This is a duplicate of #493 . The intention is we will first land this version on Aptos branch while we discuss what would be the right runtime checks to add for main branch.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/move-language/move/blob/main/CONTRIBUTING.md#developer-workflow)?
Yes

## Test Plan

`cargo test`

### CODE REVIEW

**Comment by wrwg** [comment]
Please change the target where you want to merge into to `aptos`. There should be dropdown at the top of your PR. Once you do the diff will be against that branch.

**Comment by wrwg** [comment]
@chen-robert FYI

**Comment by runtian-zhou** [comment]
The remaining work items are:
1. Get benchmarks from @vgao1996 
2. Force annotating the return value of native functions.

**Comment by chen-robert** [comment]
@runtian-zhou some additional ideas for checks which I think would make sense to get merged in here
1. Function visibility check
2. Check `ty_args` against function type parameter visibility requirements
3. Hot potato `Pack` and `Unpack` tracking

File: language/move-bytecode-verifier/src/verifier.rs [hunk 1]
```diff
@@ -20,6 +20,7 @@ use move_binary_format::{
 pub struct VerifierConfig {
     pub max_loop_depth: Option<usize>,
     pub treat_friend_as_private: bool,
+    pub strict_runtime_verification: bool,
```
Comment (wrwg) [inline language/move-bytecode-verifier/src/verifier.rs:23]: As discussed on the other PR lets not call this strict... There is really an association with 'strict' vs 'non-strict' which is a misfit here.

The name suggested by Sam, `paranoid_checks`, is much better and reflects what this feature is really about in the overall Move architecture.

File: language/move-vm/runtime/src/interpreter.rs [hunk 2]
```diff
@@ -67,6 +67,8 @@ pub(crate) struct Interpreter {
     operand_stack: Stack,
     /// The stack of active functions.
     call_stack: CallStack,
+    /// Whether to perform runtime type safety checks.
+    check_runtime_types: bool,
```
Comment (wrwg) [inline language/move-vm/runtime/src/interpreter.rs:71]: Should use the same name as above. Also it is not just type checks.

**Review by wrwg — COMMENTED** [review]
Looks mostly ready.

**Comment by runtian-zhou** [comment]
Right now the two tests are failing:
- language/move-prover/interpreter-testsuite/tests/concrete_check/pack_unpack.move
- language/move-prover/interpreter-testsuite/tests/concrete_check/destroy.move

IMHO the failure here are as expected because the test function is returning struct has zero capabilities and thus causing the struct to be dropped unexpectedly...

File: language/move-vm/runtime/src/interpreter.rs [hunk 3]
```diff
@@ -148,6 +148,17 @@ impl Interpreter {
         ty_args: Vec<Type>,
         args: Vec<Value>,
     ) -> VMResult<Vec<Value>> {
+        // A counter for tracking numbers of struct that don't have any capability.
+        // This counter will be incremented whenever a struct that has empty capability was created by pack and decremented.
+        //
```
Comment (wrwg) [inline language/move-vm/runtime/src/interpreter.rs:153]: ... decremented on unpack

File: language/move-vm/runtime/src/interpreter.rs [hunk 4]
```diff
@@ -148,6 +148,17 @@ impl Interpreter {
...
+        // The invariant here is this counter should always goes to zero at the end of execution.
+        // The reason is that values with such capability can only be created by pack instruction and destructed by unpack
+        // instruction and can never be stored in storage.
+        //
+        // Thus if the counter is not zero towards the end of execution, it means that some hot potato values are silently dropped at runtime.
+
+        let mut hot_potato_counter: u64 = 0;
```
Comment (wrwg) [inline language/move-vm/runtime/src/interpreter.rs:160]: Its fine that we do just this for now, but a more central problem of hot potatoes -- at least in the mind of the Move team when this feature was developed -- is store, not drop. Anyway good to see this.

File: language/move-vm/runtime/src/loader.rs [hunk 5]
```diff
@@ -1376,6 +1376,23 @@ impl<'a> Resolver<'a> {
         Type::Struct(struct_def)
     }
 
+    pub(crate) fn get_struct_ability(&self, idx: StructDefinitionIndex) -> AbilitySet {
```
Comment (wrwg) [inline language/move-vm/runtime/src/loader.rs:1379]: This is again very inefficient... Lets keep in mind to optimize

File: language/move-vm/runtime/src/interpreter.rs [hunk 6]
```diff
@@ -1082,6 +1114,14 @@ impl Frame {
                             interpreter.operand_stack.last_n(field_count as usize)?,
                         )?;
                         let args = interpreter.operand_stack.popn(field_count)?;
+
+                        if self.paranoid_checks {
```
Comment (wrwg) [inline language/move-vm/runtime/src/interpreter.rs:1118]: Wondering whether we should have a different flag, so if this starts to produce false positives (the risk here higher as with type checks), we don't have to turn everthing off.

**Review by wrwg — COMMENTED** [review]
For hot potato checking

File: language/move-vm/runtime/src/interpreter.rs (hunk 4 above)
Comment (runtian-zhou) [reply language/move-vm/runtime/src/interpreter.rs:160]: Key and copy can both break the invariant here: struct has key can be stored directly into storage, struct has copy can be copied and thus causing more unpack than pack.

File: language/move-vm/runtime/src/loader.rs (hunk 5 above)
Comment (runtian-zhou) [reply language/move-vm/runtime/src/loader.rs:1379]: This doesn't look very inefficient to me? Just reading out a value from a vector?

**Review by wrwg — APPROVED** [review]

File: language/move-vm/runtime/src/interpreter.rs (hunk 6 above)
Comment (runtian-zhou) [reply language/move-vm/runtime/src/interpreter.rs]: Added a different flag for this.

**Comment by jolestar** [comment]
I like this solution more than #493, I think this solution will make me sleep a little more soundly😂. Expect it to be merged into the main branch.

**Comment by sblackshear** [comment]
To expand on the issue @tnowacki was mentioning on the other PR with an example:

```
let bad_signer = ... // turn bytes into a signer via exploiting a vulnerable instruction (e.g., VecPack). This will still have a vector<u8> type tag, so the attacker must use it carefully
let vec = vector::empty();
vector::push_back(&mut vec, bad_signer); // note: this is a VecPushBack bytecode instruction, not a function call--no checks are performed
let any = any::pack(vec) // this will check the type of `vec`, but not `vec[0]`
let laundered_vec = any::unpack(any); // now vec[0] has the `signer` tag, can be used freely
```

**Comment by runtian-zhou** [comment]
> To expand on the issue @tnowacki was mentioning on the other PR with an example:
> 
> ```move
> let bad_signer = ... // turn bytes into a signer via exploiting a vulnerable instruction (e.g., VecPack). This will still have a vector<u8> type tag, so the attacker must use it carefully
> let vec = vector::empty();
> vector::push_back(&mut vec, bad_signer); // note: this is a VecPushBack bytecode instruction, not a function call--no checks are performed
> let any = any::pack(vec) // this will check the type of `vec`, but not `vec[0]`
> let laundered_vec = any::unpack(any); // now vec[0] has the `signer` tag, can be used freely
> ```

So if `any` is the function in `aptos-framework` it would be fine because we check the inner fields of struct recursively at serialization time.

If we check all type matches the declaration of Pack, we would be able to argue that in the MoveVM with paranoid mode on, no malformed value can ever be created. However, due to the limited timeline that we have I intensionally cut that corner but I would still argue that this would offer us enough safety. In this version, a module can create a struct that has bad field elements in it. But there's no way to exploit from this fact from a practical standpoint given that we check the struct types to match at function call as well.

**Comment by wrwg** [comment]
What is going on with this PR? I thought it was already ready to land the day before yesterday. @davidiw

File: language/move-vm/types/src/values/values_impl.rs [hunk 7]
```diff
@@ -1811,6 +1945,164 @@ impl Vector {
...
+            | (ValueImpl::U64(_), Type::U64)
+            | (ValueImpl::Bool(_), Type::Bool)
+            | (ValueImpl::Address(_), Type::Address) => Ok(()),
+            (ValueImpl::IndexedRef(r), Type::Reference(ty) | Type::MutableReference(ty)) => {
+                r.check_type(ty)
+            }
+            (ValueImpl::ContainerRef(r), Type::Reference(ty) | Type::MutableReference(ty)) => {
+                r.container().check_type(ty)
+            }
+            (ValueImpl::Container(c), ty) => c.check_type(ty),
```
Comment (vgao1996) [inline language/move-vm/types/src/values/values_impl.rs:1998]: Is this correct? Why are we shadowing `ty` here?

**Review by vgao1996 — APPROVED** [review]
Stamping to get us unblocked, but I share similar concerns with @sblackshear that this only seems to cover some of the type safety invariants and it's a bit hard to reason what exact subset being covered is

**Pull request merged**